### PR TITLE
validating batch table

### PR DESCRIPTION
**Pull request by JudyWeng**
#32 
I think I have the helper methods to extract and validate the batch table, but the loadJson method from Cesium returns a promise and I'm still working on learning about promises and how to use it in the individual tile validation files.

### CODE REVIEW

**Comment by pjcozzi** [comment]
@lilleyse will you be able to look at this soon?

**Comment by lilleyse** [comment]
If I remember correctly, the main things to address here were:

* Move the batch table JSON and batch table binary extraction to the individual `validateB3dm`, `validateI3dm`, files. While there will be some duplicate code, I think it is worth it just so each tile type is self-contained.
* Avoid loading the batch table schema in `validateBatchTable`, since it would be much nicer to avoid promises entirely here. Rather it could be loaded right from the start in in `validateTileset` and eventually passed along to each tile in some sort of `schemas` object.

**Comment by pjcozzi** [comment]
@JudyWeng any update here as well?

**Comment by JudyWeng** [comment]
I had attempted the method of changing the tiles to promises first, which caused a lot of changes, but I'm going to try loading the batch table schema in validateTileset now instead to avoid having to change every validateB3dm, validateI3dm file.

File: validator/specs/lib/validateB3dmSpec.js [hunk 1]
```diff
@@ -1,9 +1,27 @@
...
 
+var loadJson = Cesium.loadJson;
+
 describe('validateB3dm', function() {
+    var batchSchema;
+    beforeAll(function(done) {
+        console.log("beforeAll");
 
+        loadJson('https://raw.githubusercontent.com/AnalyticalGraphicsInc/3d-tiles/master/schema/batchTable.schema.json')
+            .then(function(schema) {
```
Comment (JudyWeng) [inline validator/specs/lib/validateB3dmSpec.js]: I tried to change the validate files to take the batch table schema in as a parameter to avoid having to change everything  into an asynchronous process, as suggested previously. However, it seems like loadJson() is returning the "XMLHttpRequest i undefined" error. I have tried both the url above and this url: "https://github.com/AnalyticalGraphicsInc/3d-tiles/blob/master/schema/batchTable.schema.json" to load the Json. Am I not using this function properly?

Comment (lilleyse) [inline validator/specs/lib/validateB3dmSpec.js]: Hm, `XMLHttpRequest` must only be supported in browsers and not Node. You could try looking at [http.request](https://nodejs.org/api/http.html#http_http_request_options_callback).

Comment (JudyWeng) [inline validator/specs/lib/validateB3dmSpec.js]: If I choose to use http.request it would require me to change loadJson() though wouldn't it?

Comment (pjcozzi) [inline validator/specs/lib/validateB3dmSpec.js]: @lilleyse any thoughts here?

Comment (lilleyse) [inline validator/specs/lib/validateB3dmSpec.js]: We discussed this offline, that using http.request (or alternatives) is fine to replace loadJson.

**Comment by JudyWeng** [comment]
I wrote the test cases for the binary table validation for b3dm, i3dm, and pnts, but not cmpt yet because that validation branch has yet to be merged into validator-master. Please let me know if there are more edge cases I have yet to cover.

**Comment by lilleyse** [comment]
I'll look at both later today.

File: validator/lib/validateB3dm.js [hunk 2]
```diff
@@ -19,7 +21,14 @@ function validateB3dm(content) {
     }
 
     if (!Buffer.isBuffer(content)) {
-        throw new DeveloperError('content must be of type buffer');
+        throw new DeveloperError('b3dm content must be of type buffer');
+    }
+
+    if(content.length < 2) {
```
Comment (lilleyse) [inline validator/lib/validateB3dm.js]: Should this be 24?

Comment (lilleyse) [inline validator/lib/validateB3dm.js]: Style: `if(` -> `if (`.

Fix throughout.

File: validator/lib/validateB3dm.js [hunk 3]
```diff
@@ -29,27 +38,90 @@ function validateB3dm(content) {
...
+
+    if(message === '') {
+        message = 'successfully extracted batch table'
+    }
+
+    return {
+        batchTableJSON: batchTableJSON,
+        batchTableBinary: batchTableBinary,
+        message: message
+    };
```
Comment (lilleyse) [inline validator/lib/validateB3dm.js]: This section may be able to be condensed by extracting the sections, and then verifying they are correct.

For example - https://github.com/AnalyticalGraphicsInc/3d-tiles-tools/blob/master/tools/lib/extractB3dm.js#L44-L46. You could do something like that and then verify the buffers are the correct size. Ideally the batch table extraction/validation could even go inline with the rest of the b3dm checking.

File: validator/lib/validateBatchTable.js [hunk 4]
```diff
@@ -0,0 +1,29 @@
+'use strict';
+var Cesium = require('cesium');
+var Ajv = require('ajv');
```
Comment (lilleyse) [inline validator/lib/validateBatchTable.js]: Add requires in alphabetical order.

File: validator/lib/validateBatchTable.js [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+module.exports = validateBatchTable;
+
+/**
+ * Checks if provided buffers follow the batch table schema
+ *
+ * @param {Object} schema - A JSON object containing the schema for the batch table.
+ * @param {Buffer} batchTableJSON - A buffer containing the JSON Header of a batch table
```
Comment (lilleyse) [inline validator/lib/validateBatchTable.js]: `JSON Header of a batch table` -> `batch table JSON`. It's not a buffer, right?

File: validator/lib/validateBatchTable.js [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+ * @param {Buffer} batchTableBinary - A buffer containing the binary body of a batch table
```
Comment (lilleyse) [inline validator/lib/validateBatchTable.js]: `binary body of a batch table` -> `batch table binary` is simpler.

File: validator/lib/validateBatchTable.js [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+ * @returns {Object} An object with two parameters - (1) a boolean for whether the batch table follows the schema
+ *                                                   (2) a message to indicate the validation result
+ */
+function validateBatchTable(schema, batchTableJSON, batchTableBinary) {
+    var ajv = new Ajv();
+    var validation = ajv.validate(schema, batchTableJSON);
```
Comment (lilleyse) [inline validator/lib/validateBatchTable.js]: `valid` instead of `validation`?

File: validator/specs/lib/validateB3dmSpec.js [hunk 8]
```diff
@@ -1,23 +1,43 @@
...
+    it('returns false if the b3dm has wrong byteLength', function() {
         expect(validateB3dm(createWrongByteLength()).result).toBe(false);
     });
 
-    it('returns true if b3dm tile matches spec', function() {
+    it('validates b3dm tile matches spec', function() {
         expect(validateB3dm(createB3dmTile()).result).toBe(true);
     });
+
+    it('validates b3dm tile with batch table JSON header matches spec', function() {
```
Comment (lilleyse) [inline validator/specs/lib/validateB3dmSpec.js]: The wording for these tests is awkward.

File: validator/specs/lib/validateB3dmSpec.js [hunk 9]
```diff
@@ -33,37 +53,140 @@ function createB3dmTile() {
...
-    var header = new Buffer(24);
+    var header = createB3dmTile();
     header.write('xxxx', 0); // magic
-    header.writeUInt32LE(1, 4); // version
-    header.writeUInt32LE(header.length, 8); // byteLength
-    header.writeUInt32LE(0, 12); // batchTableJSONByteLength
-    header.writeUInt32LE(0, 16); // batchTableBinaryByteLength
-    header.writeUInt32LE(0, 20); // batchLength
-
+    console.log('checking:\n' + header + '\n');
```
Comment (lilleyse) [inline validator/specs/lib/validateB3dmSpec.js]: Remove this.

File: validator/specs/lib/validateB3dmSpec.js [hunk 10]
```diff
@@ -33,37 +53,140 @@ function createB3dmTile() {
...
     return header;
+}
+
+function createB3dmBatchJson() {
+    var header = createB3dmTile();
+    var batchJSON = createValidBatchTableJSON();
+    header.writeUInt32LE(header.length + batchJSON.length, 8); // byteLength
+    header.writeUInt32LE(batchJSON.length, 12); // batchTableJSONByteLength
+
+    return Buffer.concat([header, batchJSON]);
```
Comment (lilleyse) [inline validator/specs/lib/validateB3dmSpec.js]: Better to avoid condensing the name, `batchTableJSON` while longer is easier to follow.

Comment (lilleyse) [inline validator/specs/lib/validateB3dmSpec.js]: The name of the function could also be `createB3dmWithBatchTableJson`.

Comment (lilleyse) [inline validator/specs/lib/validateB3dmSpec.js]: Also make sure to be consistent with `Json` vs `JSON`.

File: validator/specs/lib/validateB3dmSpec.js [hunk 11]
```diff
@@ -33,37 +53,140 @@ function createB3dmTile() {
...
+
+    return Buffer.concat([header, batchTable.buffer]);
+}
+
+function createValidBatchTableJSON() {
+    var batchJson = {
+        "id":[0,1,2],
+        "longitude":[-1.3196595204101946,-1.3196567190670823,-1.3196687138763508],
+        "height":[8,14,14]
+    };
```
Comment (lilleyse) [inline validator/specs/lib/validateB3dmSpec.js]: No need to wrap each key in quotes.

File: validator/specs/lib/validateB3dmSpec.js [hunk 12]
```diff
@@ -33,37 +53,140 @@ function createB3dmTile() {
...
+}
+
+function createValidBatchTableBinary() {
+    var batchJson = {
+        "id" : [0, 1, 2],
+        "longitude" :[-1.3196595204101946,-1.3196567190670823,-1.3196687138763508],
+        "height" : {
+            "byteOffset" : 12,
+            "componentType" : "UNSIGNED_INT",
+            "type" : "SCALAR"
```
Comment (lilleyse) [inline validator/specs/lib/validateB3dmSpec.js]: Use single quotes for `UNSIGNED_INT` and `SCALAR`.

File: validator/lib/validateBatchTable.js [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+    'VEC2': 2,
+    'VEC3': 3,
+    'VEC4': 4
+};
+
+/**
+ * Checks if provided buffers follow the batch table schema
+ *
+ * @param {Object} schema - A JSON object containing the schema for the batch table.
+ * @param {JSON} batchTableJSON - Batch table JSON
```
Comment (lilleyse) [inline validator/lib/validateBatchTable.js]: I don't think this is a valid jsdoc type, use `Object` instead.

File: validator/lib/validateBatchTable.js [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+            var binaryBodyLength = batchTableBinary.length;
+            var batchLength = batchTableJSON["id"].length;
+            var totalOffset = 0;
+            for (var key in batchTableJSON) {
+                if (!batchTableJSON.hasOwnProperty(key)) {
+                    continue;
+                }
+
+                var val = batchTableJSON[key];
+                if (Object.prototype.toString.call(val) === '[object Array]') {
```
Comment (lilleyse) [inline validator/lib/validateBatchTable.js]: Also validate that the array length equals batch length.

File: validator/lib/validateBatchTable.js [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+        if(defined(batchTableBinary) && (batchTableBinary.length > 0)) {
+            var binaryBodyLength = batchTableBinary.length;
+            var batchLength = batchTableJSON["id"].length;
+            var totalOffset = 0;
+            for (var key in batchTableJSON) {
+                if (!batchTableJSON.hasOwnProperty(key)) {
+                    continue;
+                }
+
+                var val = batchTableJSON[key];
```
Comment (lilleyse) [inline validator/lib/validateBatchTable.js]: I feel like `property` is a more fitting name.

File: validator/lib/validateBatchTable.js [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+function validateBatchTable(schema, batchTableJSON, batchTableBinary) {
+    var ajv = new Ajv();
+    var validSchema = ajv.validate(schema, batchTableJSON);
+    var valid = validSchema;
+    var message;
+
+    if(validSchema) {
+        if(defined(batchTableBinary) && (batchTableBinary.length > 0)) {
+            var binaryBodyLength = batchTableBinary.length;
+            var batchLength = batchTableJSON["id"].length;
```
Comment (lilleyse) [inline validator/lib/validateBatchTable.js]: The property `id` is not guaranteed to exist. The batch length is now in the feature table: https://github.com/AnalyticalGraphicsInc/3d-tiles/tree/master/TileFormats/Batched3DModel#global-semantics

File: validator/lib/validateI3dm.js [hunk 17]
```diff
@@ -19,7 +21,14 @@ function validateI3dm(content) {
...
 
     if (!Buffer.isBuffer(content)) {
-        throw new DeveloperError('content must be of type buffer');
+        throw new DeveloperError('i3dm content must be of type buffer');
+    }
+
+    if (content.length < 32) {
+        return {
+            result : false,
+            message: 'i3dm tile header must be 28 bytes'
```
Comment (lilleyse) [inline validator/lib/validateI3dm.js]: 32

File: validator/lib/validateI3dm.js [hunk 18]
```diff
@@ -1,5 +1,7 @@
 'use strict';
 var Cesium = require('cesium');
+var batchTableSchema = require('../specs/data/schema/batchTable.schema.json');
```
Comment (lilleyse) [inline validator/lib/validateI3dm.js]: Overall same comments here as `validateB3dm`.

File: validator/lib/validateI3dm.js [hunk 19]
```diff
@@ -30,31 +39,60 @@ function validateI3dm(content) {
...
             result : false,
-            message: 'Header has an invalid gltfFormat field. Expected gltfFormat = 0 or 1. Found gltfFormat = ' + gltfFormat
+            message: 'i3dm tile has an invalid gltfFormat field. Expected gltfFormat = 0 or 1. Found gltfFormat = ' + gltfFormat
         };
     }
 
+    var offset = 32;
+    var featureTableJSONByteLength = content.readUInt32LE(12);
+    var featureTableBinaryByteLength = content.readUInt32LE(16);
+    offset += featureTableJSONByteLength + featureTableBinaryByteLength;
```
Comment (lilleyse) [inline validator/lib/validateI3dm.js]: Style: move this a few lines below so it doesn't cut off the header value reads.

File: validator/specs/lib/validateBatchTableSpec.js [hunk 20]
```diff
@@ -0,0 +1,106 @@
...
+
+    return batchTableJSON;
+}
+
+function createInvalidBatchTableJSON() {
+    var batchTableJSON = {
+        id:[0],
+        longitude:[-1.3196595204101946],
+        height:8
+    };
```
Comment (lilleyse) [inline validator/specs/lib/validateBatchTableSpec.js]: For any of these, add some spacing 

`id : [0],`

File: validator/specs/lib/validateBatchTableSpec.js [hunk 21]
```diff
@@ -0,0 +1,106 @@
...
+        height : {
+            "byteOffset" : 0,
+            "componentType" : 'UNSIGNED_INT',
+            "type" : 'SCALAR'
+        },
+        width : {
+            "byteOffset" : 12,
+            "componentType" : 'UNSIGNED_INT',
+            "type" : 'SCALAR'
+        }
```
Comment (lilleyse) [inline validator/specs/lib/validateBatchTableSpec.js]: Fine to do:
```
width : {
    byteOffset : 12,
    componentType : 'UNSIGNED_INT',
    type : 'SCALAR'
}
```

File: validator/lib/validateBatchTable.js [hunk 22]
```diff
@@ -0,0 +1,89 @@
...
+                var numberOfComponents = val.type;
+
+                if (byteOffset < totalOffset) {
+                    valid = false;
+                    message = 'batch table property ' + key + ' has offset within another property\'s range at ' + byteOffset;
+                    break;
+                }
+
+                var propertyByteLength = batchLength * componentTypeSize[componentType] * typeSize[numberOfComponents];
+                if (propertyByteLength + byteOffset > binaryBodyLength) {
```
Comment (lilleyse) [inline validator/lib/validateBatchTable.js]: Also worth validating that `componentType` and `type` are actually correct.

File: validator/lib/validateBatchTable.js [hunk 23]
```diff
@@ -0,0 +1,89 @@
...
+    'VEC3': 3,
+    'VEC4': 4
+};
+
+/**
+ * Checks if provided buffers follow the batch table schema
+ *
+ * @param {Object} schema - A JSON object containing the schema for the batch table.
+ * @param {JSON} batchTableJSON - Batch table JSON
+ * @param {Buffer} batchTableBinary - A buffer containing the batch table binary
```
Comment (lilleyse) [inline validator/lib/validateBatchTable.js]: Since `batchTableBinary` is optional, change to `[batchTableBinary]`

File: validator/specs/lib/validateBatchTableSpec.js [hunk 24]
```diff
@@ -0,0 +1,106 @@
...
+        expect(validateBatchTable(batchTableSchema, batchTableJSON).validation).toBe(false);
+    });
+
+    it('returns false if batch table JSON does not match schema with binary body', function() {
+        var batchTableJSON = createInvalidBatchTableJSONBinary();
+        var batchTableBinary = createValidBatchTableBinary();
+        expect(validateBatchTable(batchTableSchema, batchTableJSON, batchTableBinary).validation).toBe(false);
+    });
+
+    it('byteoffset out of bounds', function() {
```
Comment (lilleyse) [inline validator/specs/lib/validateBatchTableSpec.js]: Make these two slightly more descriptive - `Binary property byteOffset out of bounds`

File: validator/specs/lib/validateCmptSpec.js [hunk 25]
```diff
@@ -186,3 +199,89 @@ function createCmptInvalidCombination() {
...
+function createCmptWithBatchTable() {
+    //cmpt1[b3dm, cmpt2[cmpt3[pnts], i3dm]]
+    var innerB3dmTile = createB3dmTile();
+    var innerPntsTile = createPntsTile();
+    var innerI3dmTile = createI3dmTile();
+
+    var batchTable = createBatchTableBinary();
+    innerB3dmTile.writeUInt32LE(innerB3dmTile.length + batchTable.buffer.length, 8); // byteLength
+    innerB3dmTile.writeUInt32LE(batchTable.batchTableJSONByteLength, 12); // batchTableJSONByteLength
+    innerB3dmTile.writeUInt32LE(batchTable.batchTableBinaryByteLength, 16);
```
Comment (lilleyse) [inline validator/specs/lib/validateCmptSpec.js]: 12 and 16 are now offsets for the feature table

**Comment by JudyWeng** [comment]
I've updated the old tests to the new b3dm specs and they pass when I run them. I'm not sure what else might be causing the checks to fail...

**Comment by lilleyse** [comment]
You can ignore that, Travis is broken for this repo right now due to the directory structure.

**Comment by lilleyse** [comment]
I'm late here, but thanks for all the work @JudyWeng!
